### PR TITLE
Allow structural anonymous classes

### PR DESCRIPTION
**Pull request by odersky**
We  now allow structural dispatch on classes that extend the `scala.Selectable` trait. If a class extends `scala.reflect.Selectable` structural dispatch is resolved using Java reflection.

### CODE REVIEW

**Comment by soronpo** [comment]
Thanks for this. AFAIK, this was the last programmability gap that prevented me from porting my library to dotty. I currently don't have time to do this, but will get on that when I can.

File: docs/docs/reference/changed-features/structural-types.md [hunk 1]
```diff
@@ -123,6 +123,43 @@ New instances of `Selectable` can be defined to support means of
...
+  val range = 240
+}
+i3.range
+```
+The type of `i3` in this example is `Vehicle { val range: Int }`. Hence,
+`i3.range` is well-formed. Since the base class `Vehicle` does not define a `range` field or method, we need structural dispatch to access the `range` field of the anonymous class that initializes `id3`. Structural dispatch
+is implemented by the base trait `reflect.Selectable` of `Vehicle`, which
+defines the necessary `selectDynamic` member.
+
+`Vehicle` could also extend some other subclass of `scala.Selectable` that implements `selectDynamic` and `applyDynamic` differently. But if it does not extend a `Selectable` at all, the code would no longer typecheck:
```
Comment (johnynek) [inline docs/docs/reference/changed-features/structural-types.md:145]: here you write `scala.Selectable` but above you write `reflect.Selectable` (and also in the tests). I assume this is a typo, but perhaps `scala.Selectable` is also a thing.

Comment (smarter) [reply docs/docs/reference/changed-features/structural-types.md:145]: Yes, scala.reflect.Selectable is the java-reflection-based implementation of the scala.Selectable mechanism https://github.com/lampepfl/dotty/blob/master/library/src/scala/Selectable.scala  https://github.com/lampepfl/dotty/blob/master/library/src/scala/reflect/Selectable.scala

Comment (johnynek) [reply docs/docs/reference/changed-features/structural-types.md:145]: Got it. Thanks for the clarification. (I never use these due to performance anxiety associated with reflection, so I had forgotten about that).

**Comment by Alexey-NM** [comment]
Would  it be possible to use dynamic invocation in such cases, is it difficult to allow:  
 - to make selectDynamic inlined function
 - to implement function in macro library which generate dynamic invocation 
See also: https://docs.oracle.com/javase/7/docs/api/java/lang/invoke/CallSite.html

**Pull request merged**